recipes/cppcheck/all/conanfile.py

```diff
@@ -1,74 +1,77 @@
+    default_options = {"with_z3": "deprecated", "have_rules": True}
 
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
+    def layout(self):
+        cmake_layout(self)
 
-    @property
-    def _build_subfolder(self):
-        return "build_subfolder"
+    def configure(self):
+        if self.options.with_z3 != "deprecated":
+            self.output.warn("with_z3 option is deprecated, do not use anymore.")
 
-    def _patch_sources(self):
-        for patch in self.conan_data["patches"][self.version]:
-            tools.patch(**patch)
-        tools.replace_in_file(os.path.join(self._source_subfolder, "cli", "CMakeLists.txt"),
-                              "RUNTIME DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}",
-                              "DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}")
+    def package_id(self):
+        del self.info.options.with_z3
 
     def requirements(self):
-        if self.options.with_z3:
-            self.requires("z3/4.8.8")
         if self.options.have_rules:
             self.requires("pcre/8.45")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
-    @functools.lru_cache(1)
-    def _configure_cmake(self):
-        cmake = CMake(self)
-        cmake.definitions["USE_Z3"] = self.options.with_z3
-        cmake.definitions["HAVE_RULES"] = self.options.have_rules
-        cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
-        cmake.definitions["ENABLE_OSS_FUZZ"] = False
-        cmake.configure(build_folder=self._build_subfolder)
-        return cmake
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["HAVE_RULES"] = self.options.have_rules
+        tc.variables["USE_MATCHCOMPILER"] = "Auto"
+        tc.variables["ENABLE_OSS_FUZZ"] = False
+        tc.generate()
+
+        deps = CMakeDeps(self)
+        deps.generate()
 
     def build(self):
-        self._patch_sources()
-        cmake = self._configure_cmake()
+        cmake = CMake(self)
+        replace_in_file(self, os.path.join(self.source_folder, "tools", "CMakeLists.txt"), "CMAKE_SOURCE_DIR", "PROJECT_SOURCE_DIR")
+        replace_in_file(self, os.path.join(self.source_folder, "cmake", "findDependencies.cmake"), "PCRE_LIBRARY pcre", "PCRE_LIBRARY NAMES pcre pcred")
+        replace_in_file(self, os.path.join(self.source_folder, "htmlreport", "cppcheck-htmlreport"), "#!/usr/bin/env python", "#!/usr/bin/env python3")
+        replace_in_file(self, os.path.join(self.source_folder, "cli", "CMakeLists.txt"),
+                              "RUNTIME DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}",
+                              "DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}")
+        replace_in_file(self, os.path.join(self.source_folder, "cli", "CMakeLists.txt"), "add_dependencies(cppcheck run-dmake)", "")
```

Comment:
It's prefered to use patch files - they are eaiser to review and reason about the changes. https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/policy_patching.md#patches-format-and-conventions

It've difficult to understand the side effects like this --- I would be greatly appreciate if you put back the old patches

Reply:
I reverted the changes